internal/core/artifact.go

```diff
@@ -0,0 +1,438 @@
+package core
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/google/uuid"
+	"github.com/kubeflow/model-registry/internal/apiutils"
+	"github.com/kubeflow/model-registry/internal/converter"
+	"github.com/kubeflow/model-registry/internal/db/models"
+	"github.com/kubeflow/model-registry/internal/mapper"
+	"github.com/kubeflow/model-registry/pkg/api"
+	"github.com/kubeflow/model-registry/pkg/openapi"
+)
+
+type ModelRegistryService struct {
+	artifactRepository           models.ArtifactRepository
+	modelArtifactRepository      models.ModelArtifactRepository
+	docArtifactRepository        models.DocArtifactRepository
+	registeredModelRepository    models.RegisteredModelRepository
+	modelVersionRepository       models.ModelVersionRepository
+	servingEnvironmentRepository models.ServingEnvironmentRepository
+	inferenceServiceRepository   models.InferenceServiceRepository
+	serveModelRepository         models.ServeModelRepository
+	mapper                       mapper.EmbedMDMapper
+}
+
+func NewModelRegistryService(
+	artifactRepository models.ArtifactRepository,
+	modelArtifactRepository models.ModelArtifactRepository,
+	docArtifactRepository models.DocArtifactRepository,
+	registeredModelRepository models.RegisteredModelRepository,
+	modelVersionRepository models.ModelVersionRepository,
+	servingEnvironmentRepository models.ServingEnvironmentRepository,
+	inferenceServiceRepository models.InferenceServiceRepository,
+	serveModelRepository models.ServeModelRepository,
+	typesMap map[string]int64) *ModelRegistryService {
+	return &ModelRegistryService{
+		artifactRepository:           artifactRepository,
+		modelArtifactRepository:      modelArtifactRepository,
+		docArtifactRepository:        docArtifactRepository,
+		registeredModelRepository:    registeredModelRepository,
+		modelVersionRepository:       modelVersionRepository,
+		servingEnvironmentRepository: servingEnvironmentRepository,
+		inferenceServiceRepository:   inferenceServiceRepository,
+		serveModelRepository:         serveModelRepository,
+		mapper:                       *mapper.NewEmbedMDMapper(typesMap),
+	}
+}
+
+func (b *ModelRegistryService) upsertArtifact(artifact *openapi.Artifact, modelVersionId *string) (*openapi.Artifact, error) {
+	var modelVersionIDPtr *int32
+
+	artToReturn := &openapi.Artifact{}
+
+	if artifact == nil {
+		return nil, fmt.Errorf("invalid artifact pointer, cannot be nil: %w", api.ErrBadRequest)
+	}
+
+	// If modelVersionId is nil, we need to generate a new UUID for the model version ID for the name prefixing,
+	// the id will still be nil when invoking the Save method so that the attribution is created only when needed.
+	if modelVersionId == nil {
+		uuid := uuid.New().String()
+
+		modelVersionId = &uuid
+	} else {
+		convertedId, err := strconv.Atoi(*modelVersionId)
```

Comment:
Be careful with these `int64` to `int32` conversions. I see a few instances of this, and I'm not going to point them all out, but anywhere that the value from `Atoi` is cast to an `int32` should be corrected.

```suggestion
		convertedId, err := strconv.ParseInt(*modelVersionId, 10, 32)
```